crates/sui-framework/sources/collectible/kiosk.move

```diff
@@ -0,0 +1,564 @@
+    ): Coin<SUI> {
+        assert!(object::id(self) == cap.for, ENotOwner);
+
+        let amount = if (option::is_some(&amount)) {
+            let amt = option::destroy_some(amount);
+            assert!(amt <= balance::value(&self.profits), ENotEnough);
+            amt
+        } else {
+            balance::value(&self.profits)
+        };
+
+        coin::take(&mut self.profits, amount, ctx)
+    }
+
+    // === Kiosk fields access ===
+
+    /// Check whether the `KioskOwnerCap` matches the `Kiosk`.
+    public fun has_access(self: &mut Kiosk, cap: &KioskOwnerCap): bool {
+        object::id(self) == cap.for
+    }
+
+    /// Access the `UID` using the `KioskOwnerCap`.
+    public fun uid_mut_as_owner(self: &mut Kiosk, cap: &KioskOwnerCap): &mut UID {
+        assert!(object::id(self) == cap.for, ENotOwner);
+        &mut self.id
+    }
+
+    /// Get the mutable `UID` for dynamic field access and extensions.
+    /// Aborts if `allow_extensions` set to `false`.
+    public fun uid_mut(self: &mut Kiosk): &mut UID {
+        assert!(self.allow_extensions, EExtensionsDisabled);
+        &mut self.id
+    }
+
+    /// Get the owner of the Kiosk.
+    public fun owner(self: &Kiosk): address {
+        self.owner
+    }
+
+    /// Get the number of items stored in a Kiosk.
+    public fun item_count(self: &Kiosk): u32 {
+        self.item_count
+    }
+
+    /// Get the amount of profits collected by selling items.
+    public fun profits_amount(self: &Kiosk): u64 {
+        balance::value(&self.profits)
+    }
+
+    /// Get mutable access to `profits` - useful for extendability.
+    public fun profits_mut(self: &mut Kiosk, cap: &KioskOwnerCap): &mut Balance<SUI> {
+        assert!(object::id(self) == cap.for, ENotOwner);
+        &mut self.profits
+    }
+
+    // === PurchaseCap fields access ===
+
+    /// Get the `kiosk_id` from the `PurchaseCap`.
+    public fun purchase_cap_kiosk<T: key + store>(self: &PurchaseCap<T>): ID {
+        self.kiosk_id
+    }
+
+    /// Get the `Item_id` from the `PurchaseCap`.
+    public fun purchase_cap_item<T: key + store>(self: &PurchaseCap<T>): ID {
+        self.item_id
+    }
+
+    /// Get the `min_price` from the `PurchaseCap`.
+    public fun purchase_cap_min_price<T: key + store>(self: &PurchaseCap<T>): u64 {
+        self.min_price
+    }
+}
+
+#[test_only]
+/// Kiosk testing strategy:
+/// - [ ] test purchase flow
+/// - [ ] test purchase cap flow
+/// - [ ] test withdraw methods
+module sui::kiosk_tests {
+    use std::vector;
+    use sui::tx_context::{Self, TxContext};
+    use sui::transfer_policy::{Self as policy, TransferPolicy, TransferPolicyCap};
+    use sui::kiosk::{Self, Kiosk, KioskOwnerCap};
+    use sui::object::{Self, ID, UID};
+    use sui::sui::SUI;
+    use sui::package;
+    use sui::coin;
+
+    const AMT: u64 = 10_000;
+
+    struct Asset has key, store { id: UID }
+    struct OTW has drop {}
+
+    /// Prepare: accounts
+    /// Alice, Bob and my favorite guy - Carl
```

Comment:
Just out of interest -- llamas wearing hats?

Reply:
Caaaaarl!